presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -323,6 +319,39 @@ private synchronized void addDynamicFilters(QueryId queryId, Map<DynamicFilterId
                         descriptor -> columnHandles.get(Symbol.from(descriptor.getInput()))));
     }
 
+    private static Set<DynamicFilterId> getLazyDynamicFilters(PlanFragment plan)
+    {
+        // lazy dynamic filters cannot be consumed by the same stage where they are produced as it would result in query deadlock
+        return difference(getProducedDynamicFilters(plan.getRoot()), getConsumedDynamicFilters(plan.getRoot()));
+    }
+
+    private static Set<DynamicFilterId> getReplicatedDynamicFilters(PlanNode planNode)
+    {
```

Comment:
Would it be possible to compute the "non-lazy" dynamic filters using:
```java
intersection(getProducedDynamicFilters(plan.getRoot()), getConsumedDynamicFilters(plan.getRoot()))
```
so we'll make sure that every produced dynamic filter is handled...

Reply:
> Would it be possible to compute the "non-lazy" dynamic filters using:

what would be the difference here in practice?

I prefer that lazy DFs are selected explicitly as they can result in deadlock

Reply:
> I prefer that lazy DFs are selected explicitly as they can result in deadlock

Agree :)